File: github_cpp/16/22.cpp
#include <iostream>

using namespace std;

#define pa(a, n) do { \
	for (int i = 0; i < n; ++i) \
		cout << a[i] << " "; \
	cout << endl; \
} while(0)

void shell_sort(int* a, int n) {
	for (int gap = n; gap > 0; gap >>= 1) {
		for (int i = gap; i < n; ++i) {
			int tmp = a[i];
			int j;
			for (j = i - gap; j >= 0 && a[j] > tmp; j -= gap) {
				a[j+gap] = a[j];
			}
			a[j+gap] = tmp;
		}
	}
}

int main() {
    int a[] = {321, 32, 32, 43, 4, 65, 87, 32, 4342, 43, 5454, 6776, 564};
	const int n = sizeof(a)/sizeof(a[0]);
	pa(a, n);
    shell_sort(a, n);
	pa(a, n);

    return 0;
}

File: github_java_sort/7/22.java
public class shellsort {
	public void shellSort(int[] array) {
		for(int gap = array.length / 2; gap > 0; gap /= 2) {
			for(int i = 0; i < gap; i++) {
			    for(int j = i; j < array.length; j += gap) {
				    int min = j;
				    for(int k = j + gap; k < array.length; k += gap) {
					    if(array[k] < array[min]) {
						    min = k;
						}
					}
					swap(array, j, min);
				}
			}
		}
	}

	public void swap(int[] array, int a, int b) {
		int temp = array[a];
		array[a] = array[b];
		array[b] = temp;
	}
	public static void main(String[] args) {
		int[] array = {1,2,2,2,3, 10, 3, 40, 20, 50, 80, 70,12, 16, 22, 30, 35, 39, 42, 45, 48, 50, 53, 55, 56};
		shellsort ss = new shellsort();
		ss.shellSort(array);
		for(int i : array)
			System.out.println(i);
	}
}